File: butter.cpp
/*
ID: wrwwctb1
TASK: butter
LANG: C++14
*/

#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <cassert>

#include <vector>
#include <deque>
#include <list>
#include <set> // has multiset
#include <map> // has multimap
#include <unordered_set> // has unordered_multiset
#include <unordered_map> // has unordered_multimap
#include <stack>
#include <queue> // has priority_queue

#include <cmath>
#include <algorithm>
#include <cfloat>
#include <iomanip>
#include <ctime>
#include <climits>

#define FOR( i, N ) for( int (i) = 0; (i) < (N); (i)++ )
//#define FOR_( i, start, endplusone ) for( int (i) = (start); (i) < (endplusone); (i)++ )
const double EPS = 1e-6;
const double PI = 3.141592653589793;
using namespace std;



const string filename("butter");








int dijk(vector<vector<pair<int, int>>> &neig,
         int s,
         int P,
         vector<int> &cowlocs){

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.emplace(0, s);
    vector<bool> seen(P+1, false);
    vector<int> est(P+1, INT_MAX);
    est[s] = 0;

    int tot = 0;

    while (!pq.empty()){
        pair<int, int> est_u = pq.top();
        pq.pop();
        int estu = est_u.first;
        int u = est_u.second;

        if (seen[u])
            continue;

        seen[u] = true;

        if (cowlocs[u])
            tot += estu * cowlocs[u];

        for (auto wv: neig[u]){
            int w = wv.first;
            int v = wv.second;

            if (seen[v])
                continue;

            if (estu + w < est[v]){
                est[v] = estu + w;
                pq.emplace(estu+w, v);
            }
        }
    }
    return tot;
}

int main(int argc, char **argv){

    string filenameo = filename + ".out";
    string filenamei = filename + ".in";
    ofstream fout(filenameo.c_str());
    ifstream fin(filenamei.c_str());


    int N, P, C;
    fin >> N >> P >> C;

    // const int Nmax = 500;
    // const int Pmax = 800;
    // const int Cmax = 1450;
    vector<int> cowlocs(P+1, 0);
    FOR(n, N){
        int loc;
        fin >> loc;
        cowlocs[loc] += 1;
    }

    vector<vector<pair<int, int>>> neig(P+1);
    int u, v, w;
    FOR(c, C){
        fin >> u >> v >> w;
        neig[u].push_back(make_pair(w, v));
        neig[v].push_back(make_pair(w, u));
    }

    // // check input
    // cout << " " << N << " " << P << " " << C << endl;
    // for (auto c: cowlocs)
    //     cout << " " << c;
    // cout << endl;
    // for (auto v: neig){
    //     for (auto p: v)
    //         cout << " " << p.first << " " << p.second << " ";
    //     cout << endl;
    // }

    int best = INT_MAX;
    for (int sugar = 1; sugar <= P; sugar++){
        int tot = dijk(neig, sugar, P, cowlocs);
        if (tot < best)
            best = tot;
    }

    fout << best << endl;







    //system("pause");
    return 0;
}

File: butter.py
"""
ID: wrwwctb1
LANG: PYTHON3
TASK: butter
3.2.7

key idea
dijk from every pasture. calc tot travel dist within each dijk run

if decrease key is available: N C log P = 500 * 1450 * log 800 ~ 7e6
for this implementation     : N C log C

can simplify dijk if nodes have consecutive int names

can skip dict membership check (for neig, est)
"""

import os, sys, re
from collections import Counter, deque, defaultdict
from copy import copy
from itertools import combinations, permutations, accumulate, \
                      combinations_with_replacement
from functools import lru_cache, cmp_to_key, partial as functools_partial
from heapq import heappush, heappop, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from math import ceil, floor, factorial, gcd, modf, log, log2, log10, sqrt, \
         pi, sin, cos, tan, asin, acos, atan, atan2, hypot, erf, erfc, inf, nan
# sys.setrecursionlimit(5782)
# python -m cProfile -s time ha.py

print = functools_partial(print, flush=True)
input = lambda: fin.readline().strip('\n')

def printwrite(string):
    print(string)
    fout.write(str(string) + '\n')

filename = 'butter'
fin = open(filename + '.in', 'r')
fout = open(filename + '.out', 'w')













def dijk(neig, s, P, cowlocs):
    pq = [(0, s)]
    seen = [False] * (P+1) # was set
    est = [inf] * (P+1) # was dict
    est[s] = 0

    tot = 0

    while pq:
        estu, u = heappop(pq)
        if seen[u]:
            continue

        seen[u] = True

        if cowlocs[u]: # optimization. no need to do this outside
           tot += estu * cowlocs[u]

        for w, v in neig[u]:
            if seen[v]:
                continue

            if estu + w < est[v]:
                est[v] = cand = estu + w
                heappush(pq, (cand, v))
    return tot


#import time
#tt = time.time()

N, P, C = map(int, input().split())
cowlocs = [0] * (P+1) # cowlocs[loc] = # of cows
for n in range(N):
    loc = int(input())
    cowlocs[loc] += 1

neig = [[] for p in range(P+1)]
for c in range(C):
    u, v, w = map(int, input().split())
    neig[u].append((w, v))
    neig[v].append((w, u))


best = inf
for sugar in range(1, P+1):
    tot = dijk(neig, sugar, P, cowlocs)
    if tot < best:
        best = tot

printwrite(best)


#print(time.time()-tt)









fin.close()
fout.close()
del print, input

